pyrealm/pmodel/calc_optimal_chi_new.py

```diff
@@ -0,0 +1,573 @@
+        pmodel_const: An instance of
+            :class:`~pyrealm.constants.pmodel_const.PModelConst`.
+
+    Returns:
+        Instances of the abstract base class should not be created - use instances of
+        specific subclasses.
+    """
+
+    # @property
+    # @abstractmethod
+    # def method(cls) -> str:
+    #     """The method name.
+
+    #     This class property sets the name used to refer to identify the class in
+    #     the :data:`~pyrealm.pmodel.calc_optimal_chi.OPTIMAL_CHI_METHOD_REGISTRY`.
+    #     """
+
+    # @property
+    # @abstractmethod
+    # def is_c4(cls) -> bool:
+    #     """Does the class represent a C4 pathway."""
+
+    method: str
+    is_c4: bool
+
+    def __init__(
+        self,
+        env: PModelEnvironment,
+        rootzonestress: NDArray = np.array([1.0]),
+        pmodel_const: PModelConst = PModelConst(),
+    ):
+        self.env: PModelEnvironment = env
+        """The PModelEnvironment containing the photosynthetic environment for the
+        model."""
+
+        # If rootzonestress is not simply equal to 1 (or an equivalent ndarray), check
+        # rootzonestress conforms to the environment data
+        if np.allclose(rootzonestress, 1.0):
+            self.shape: tuple = env.shape
+            """Records the common numpy array shape of array inputs."""
+        else:
+            self.shape = check_input_shapes(env.ca, rootzonestress)
+            warn("The rootzonestress option is an experimental feature.")
+
+        self.rootzonestress: NDArray = rootzonestress
+        """Experimental rootzonestress factor, unitless."""
+
+        self.pmodel_const: PModelConst = pmodel_const
+        """The PModelParams used for optimal chi estimation"""
+
+        # Declare attributes populated by methods. These are typed but not assigned a
+        # default value as they must be populated by the set_beta and estimate_chi
+        # methods, which are called below, and so will be populated before __init__
+        # returns.
+        self.beta: NDArray
+        """The ratio of carboxylation to transpiration cost factors."""
+        self.xi: NDArray
+        r"""Defines the sensitivity of :math:`\chi` to the vapour pressure deficit,
+        related to the carbon cost of water (Medlyn et al. 2011; Prentice et 2014)."""
+        self.chi: NDArray
+        r"""The ratio of leaf internal to ambient :math:`\ce{CO2}` partial pressure
+        (:math:`\chi`)."""
+        self.ci: NDArray
+        r"""The leaf internal :math:`\ce{CO2}` partial pressure (:math:`c_i`)."""
+        self.mc: NDArray
+        r""":math:`\ce{CO2}` limitation factor for RuBisCO-limited assimilation
+        (:math:`m_c`)."""
+        self.mj: NDArray
+        r""":math:`\ce{CO2}` limitation factor for light-limited assimilation
+        (:math:`m_j`)."""
+        self.mjoc: NDArray
+        r"""Ratio of :math:`m_j/m_c`."""
+
+        self.set_beta()
+        self.estimate_chi()
+
```

Comment:
Maybe add a check to ensure that any custom subclass of `NewCalcOptimalChi` sets the values of `beta`, `xi`, etc. in its `set_beta()` and `estimate_chi()` methods? Besides, it seems that some of these variables are derived from the others, so maybe they can be set after calling `self.estimate_chi()`. For example,
```python
self.ci = self.chi * self.env.ca
self.mjoc = self.mj / self.mc
```

Reply:
It is unlike me not to have added that checking - I'm usually pretty defensive in coding. I guess my thinking here is that a user might pass in bad inputs by mistake, but someone developing a new class is probably going to take more care. We could simply run check shape again though.

It is true that `ci` and `mjoc` are shared - I'd left them there in case a method came along that needed to define them differently.  That seems very unlikely though.

Reply:
Yes, adding a shape check is a good idea, ensuring both existence of these attributes and the consistency of their shapes.